diff --git a/libcpu/aarch64/common/gicv3.c b/libcpu/aarch64/common/gicv3.c
--- a/libcpu/aarch64/common/gicv3.c
+++ b/libcpu/aarch64/common/gicv3.c
@@ -117,23 +117,15 @@ rt_uint64_t arm_gic_get_pending_irq(rt_uint64_t index, int irq)
     irq = irq - _gic_table[index].offset;
     RT_ASSERT(irq >= 0);

-    if (irq >= 16)
+    if (irq < 32)
     {
-        pend = (GIC_DIST_PENDING_SET(_gic_table[index].dist_hw_base, irq) >> (irq % 32)) & 0x1;
+        rt_int32_t cpu_id = rt_hw_cpu_id();
+
+        pend = (GIC_RDISTSGI_ISPENDR0(_gic_table[index].redist_hw_base[cpu_id]) >> irq) & 0x1;
     }
     else
     {
-        /* INTID 0-15 Software Generated Interrupt */
-        pend = (GIC_DIST_SPENDSGI(_gic_table[index].dist_hw_base, irq) >> ((irq % 4) * 8)) & 0xff;
-        /* No CPU identification offered */
-        if (pend != 0)
-        {
-            pend = 1;
-        }
-        else
-        {
-            pend = 0;
-        }
+        pend = (GIC_DIST_PENDING_SET(_gic_table[index].dist_hw_base, irq) >> (irq % 32)) & 0x1;
     }

     return pend;
@@ -141,20 +133,23 @@ rt_uint64_t arm_gic_get_pending_irq(rt_uint64_t index, int irq)

 void arm_gic_set_pending_irq(rt_uint64_t index, int irq)
 {
+    rt_uint64_t mask;
+
     RT_ASSERT(index < ARM_GIC_MAX_NR);

     irq = irq - _gic_table[index].offset;
     RT_ASSERT(irq >= 0);
+    mask = 1UL << (irq % 32);

-    if (irq >= 16)
+    if (irq < 32)
     {
-        GIC_DIST_PENDING_SET(_gic_table[index].dist_hw_base, irq) = 1 << (irq % 32);
+        rt_int32_t cpu_id = rt_hw_cpu_id();
+
+        GIC_RDISTSGI_ISPENDR0(_gic_table[index].redist_hw_base[cpu_id]) = mask;
     }
     else
     {
-        /* INTID 0-15 Software Generated Interrupt */
-        /* Forward the interrupt to the CPU interface that requested it */
-        GIC_DIST_SOFTINT(_gic_table[index].dist_hw_base) = (irq | 0x02000000);
+        GIC_DIST_PENDING_SET(_gic_table[index].dist_hw_base, irq) = mask;
     }
 }

@@ -166,15 +161,16 @@ void arm_gic_clear_pending_irq(rt_uint64_t index, int irq)

     irq = irq - _gic_table[index].offset;
     RT_ASSERT(irq >= 0);
+    mask = 1UL << (irq % 32);

-    if (irq >= 16)
+    if (irq < 32)
     {
-        mask = 1 << (irq % 32);
-        GIC_DIST_PENDING_CLEAR(_gic_table[index].dist_hw_base, irq) = mask;
+        rt_int32_t cpu_id = rt_hw_cpu_id();
+
+        GIC_RDISTSGI_ICPENDR0(_gic_table[index].redist_hw_base[cpu_id]) = mask;
     }
     else
     {
-        mask = 1 << ((irq % 4) * 8);
-        GIC_DIST_CPENDSGI(_gic_table[index].dist_hw_base, irq) = mask;
+        GIC_DIST_PENDING_CLEAR(_gic_table[index].dist_hw_base, irq) = mask;
     }
 }
