diff --git a/libcpu/aarch64/common/gic.c b/libcpu/aarch64/common/gic.c
--- a/libcpu/aarch64/common/gic.c
+++ b/libcpu/aarch64/common/gic.c
@@ -108,6 +108,19 @@ void arm_gic_umask(rt_uint64_t index, int irq)
     GIC_DIST_ENABLE_SET(_gic_table[index].dist_hw_base, irq) = mask;
 }
 
+rt_uint64_t arm_gic_get_enable_irq(rt_uint64_t index, int irq)
+{
+    rt_uint64_t enabled;
+
+    RT_ASSERT(index < ARM_GIC_MAX_NR);
+
+    irq = irq - _gic_table[index].offset;
+    RT_ASSERT(irq >= 0U);
+
+    enabled = GIC_DIST_ENABLE_SET(_gic_table[index].dist_hw_base, irq);
+    return (enabled >> (irq % 32U)) & 0x1UL;
+}
+
 rt_uint64_t arm_gic_get_pending_irq(rt_uint64_t index, int irq)
 {
     rt_uint64_t pend;
diff --git a/libcpu/aarch64/common/gicv3.c b/libcpu/aarch64/common/gicv3.c
--- a/libcpu/aarch64/common/gicv3.c
+++ b/libcpu/aarch64/common/gicv3.c
@@ -109,6 +109,29 @@ void arm_gic_umask(rt_uint64_t index, int irq)
     }
 }
 
+rt_uint64_t arm_gic_get_enable_irq(rt_uint64_t index, int irq)
+{
+    rt_uint64_t enabled;
+
+    RT_ASSERT(index < ARM_GIC_MAX_NR);
+
+    irq = irq - _gic_table[index].offset;
+    RT_ASSERT(irq >= 0);
+
+    if (irq < 32)
+    {
+        rt_int32_t cpu_id = rt_hw_cpu_id();
+
+        enabled = GIC_RDISTSGI_ISENABLER0(_gic_table[index].redist_hw_base[cpu_id]);
+    }
+    else
+    {
+        enabled = GIC_DIST_ENABLE_SET(_gic_table[index].dist_hw_base, irq);
+    }
+
+    return (enabled >> (irq % 32)) & 0x1;
+}
+
 rt_uint64_t arm_gic_get_pending_irq(rt_uint64_t index, int irq)
 {
     rt_uint64_t pend;
diff --git a/libcpu/aarch64/common/include/gic.h b/libcpu/aarch64/common/include/gic.h
--- a/libcpu/aarch64/common/include/gic.h
+++ b/libcpu/aarch64/common/include/gic.h
@@ -20,6 +20,7 @@ void arm_gic_ack(rt_uint64_t index, int irq);
 
 void arm_gic_mask(rt_uint64_t index, int irq);
 void arm_gic_umask(rt_uint64_t index, int irq);
+rt_uint64_t arm_gic_get_enable_irq(rt_uint64_t index, int irq);
 
 rt_uint64_t arm_gic_get_pending_irq(rt_uint64_t index, int irq);
 void arm_gic_set_pending_irq(rt_uint64_t index, int irq);
diff --git a/libcpu/aarch64/common/include/gicv3.h b/libcpu/aarch64/common/include/gicv3.h
--- a/libcpu/aarch64/common/include/gicv3.h
+++ b/libcpu/aarch64/common/include/gicv3.h
@@ -145,6 +145,7 @@ void arm_gic_ack(rt_uint64_t index, int irq);
 
 void arm_gic_mask(rt_uint64_t index, int irq);
 void arm_gic_umask(rt_uint64_t index, int irq);
+rt_uint64_t arm_gic_get_enable_irq(rt_uint64_t index, int irq);
 
 rt_uint64_t arm_gic_get_pending_irq(rt_uint64_t index, int irq);
 void arm_gic_set_pending_irq(rt_uint64_t index, int irq);
diff --git a/libcpu/aarch64/common/include/interrupt.h b/libcpu/aarch64/common/include/interrupt.h
--- a/libcpu/aarch64/common/include/interrupt.h
+++ b/libcpu/aarch64/common/include/interrupt.h
@@ -26,6 +26,9 @@ void rt_hw_vector_init(void);
 void rt_hw_interrupt_init(void);
 void rt_hw_interrupt_mask(int vector);
 void rt_hw_interrupt_umask(int vector);
+#ifndef SOC_BCM283x
+unsigned int rt_hw_interrupt_get_enable(int vector);
+#endif
 
 int rt_hw_interrupt_get_irq(void);
 void rt_hw_interrupt_ack(int vector);
diff --git a/libcpu/aarch64/common/interrupt.c b/libcpu/aarch64/common/interrupt.c
--- a/libcpu/aarch64/common/interrupt.c
+++ b/libcpu/aarch64/common/interrupt.c
@@ -170,6 +170,18 @@ if (vector < 32)
 #endif
 }
 
+#ifndef SOC_BCM283x
+/**
+ * This function returns whether an interrupt is enabled on this CPU.
+ * @param vector the interrupt number
+ * @return 0 if disabled; 1 if enabled
+ */
+unsigned int rt_hw_interrupt_get_enable(int vector)
+{
+    return (unsigned int)arm_gic_get_enable_irq(0, vector);
+}
+#endif
+
 /**
  * This function returns the active interrupt number.
  * @param none
